File: archive/pulumi-py/Pr6231/lib/ecs.py
"""
ECS Cluster and Service
Creates ECS Fargate cluster and service with task definition
"""

import pulumi
import pulumi_aws as aws
import json

def create_ecs_cluster(environment_suffix: str):
    """
    Create ECS Fargate cluster
    """

    cluster = aws.ecs.Cluster(
        f"flask-cluster-{environment_suffix}",
        name=f"flask-cluster-{environment_suffix}",
        settings=[
            aws.ecs.ClusterSettingArgs(
                name="containerInsights",
                value="enabled"
            )
        ],
        tags={
            "Name": f"flask-cluster-{environment_suffix}",
            "EnvironmentSuffix": environment_suffix
        }
    )

    return cluster

def create_ecs_service(  # pylint: disable=too-many-positional-arguments,too-many-arguments
        environment_suffix: str,
        cluster,
        private_subnets,
        security_group,
        target_group,
        ecr_repo,
        db_secret,
        listener=None):
    """
    Create ECS service with Fargate tasks
    """

    # Create CloudWatch log group
    log_group = aws.cloudwatch.LogGroup(
        f"ecs-flask-{environment_suffix}",
        name=f"/ecs/flask-{environment_suffix}",
        retention_in_days=7,
        tags={
            "Name": f"ecs-flask-{environment_suffix}",
            "EnvironmentSuffix": environment_suffix
        }
    )

    # Create IAM role for task execution
    task_execution_role = aws.iam.Role(
        f"ecs-task-execution-role-{environment_suffix}",
        name=f"ecs-task-execution-role-{environment_suffix}",
        assume_role_policy=json.dumps({
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Principal": {
                        "Service": "ecs-tasks.amazonaws.com"
                    },
                    "Action": "sts:AssumeRole"
                }
            ]
        }),
        tags={
            "Name": f"ecs-task-execution-role-{environment_suffix}",
            "EnvironmentSuffix": environment_suffix
        }
    )

    # Attach AWS managed policy for ECS task execution
    task_execution_policy_attachment = aws.iam.RolePolicyAttachment(
        f"ecs-task-execution-policy-{environment_suffix}",
        role=task_execution_role.name,
        policy_arn="arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy"
    )

    # Create inline policy for Secrets Manager access
    secrets_policy = aws.iam.RolePolicy(
        f"ecs-secrets-policy-{environment_suffix}",
        role=task_execution_role.id,
        policy=db_secret.arn.apply(lambda arn: json.dumps({
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Action": [
                        "secretsmanager:GetSecretValue"
                    ],
                    "Resource": arn
                }
            ]
        }))
    )

    # Create IAM role for task
    task_role = aws.iam.Role(
        f"ecs-task-role-{environment_suffix}",
        name=f"ecs-task-role-{environment_suffix}",
        assume_role_policy=json.dumps({
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Principal": {
                        "Service": "ecs-tasks.amazonaws.com"
                    },
                    "Action": "sts:AssumeRole"
                }
            ]
        }),
        tags={
            "Name": f"ecs-task-role-{environment_suffix}",
            "EnvironmentSuffix": environment_suffix
        }
    )

    # Add policy for DynamoDB access
    dynamodb_policy = aws.iam.RolePolicy(
        f"ecs-dynamodb-policy-{environment_suffix}",
        role=task_role.id,
        policy=json.dumps({
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Action": [
                        "dynamodb:PutItem",
                        "dynamodb:GetItem",
                        "dynamodb:UpdateItem",
                        "dynamodb:DeleteItem",
                        "dynamodb:Query",
                        "dynamodb:Scan"
                    ],
                    "Resource": f"arn:aws:dynamodb:*:*:table/user-sessions-{environment_suffix}"
                }
            ]
        })
    )

    # Create task definition
    task_definition = aws.ecs.TaskDefinition(
        f"flask-task-{environment_suffix}",
        family=f"flask-task-{environment_suffix}",
        network_mode="awsvpc",
        requires_compatibilities=["FARGATE"],
        cpu="512",
        memory="1024",
        execution_role_arn=task_execution_role.arn,
        task_role_arn=task_role.arn,
        container_definitions=pulumi.Output.all(ecr_repo.repository_url, log_group.name, db_secret.arn).apply(
            lambda args: json.dumps([
                {
                    "name": "flask-app",
                    "image": f"{args[0]}:latest",
                    "essential": True,
                    "portMappings": [
                        {
                            "containerPort": 5000,
                            "protocol": "tcp"
                        }
                    ],
                    "environment": [
                        {
                            "name": "FLASK_ENV",
                            "value": "production"
                        },
                        {
                            "name": "DYNAMODB_TABLE",
                            "value": f"user-sessions-{environment_suffix}"
                        }
                    ],
                    "secrets": [
                        {
                            "name": "DATABASE_URL",
                            "valueFrom": args[2]
                        }
                    ],
                    "logConfiguration": {
                        "logDriver": "awslogs",
                        "options": {
                            "awslogs-group": args[1],
                            "awslogs-region": "us-east-2",
                            "awslogs-stream-prefix": "flask"
                        }
                    },
                    "healthCheck": {
                        "command": ["CMD-SHELL", "curl -f http://localhost:5000/health || exit 1"],
                        "interval": 30,
                        "timeout": 5,
                        "retries": 3,
                        "startPeriod": 60
                    }
                }
            ])
        ),
        tags={
            "Name": f"flask-task-{environment_suffix}",
            "EnvironmentSuffix": environment_suffix
        }
    )

    # Create ECS service
    depends_on_resources = [target_group]
    if listener:
        depends_on_resources.append(listener)

    service = aws.ecs.Service(
        f"flask-service-{environment_suffix}",
        name=f"flask-service-{environment_suffix}",
        cluster=cluster.arn,
        task_definition=task_definition.arn,
        desired_count=2,
        launch_type="FARGATE",
        network_configuration=aws.ecs.ServiceNetworkConfigurationArgs(
            subnets=[subnet.id for subnet in private_subnets],
            security_groups=[security_group.id],
            assign_public_ip=False
        ),
        load_balancers=[
            aws.ecs.ServiceLoadBalancerArgs(
                target_group_arn=target_group.arn,
                container_name="flask-app",
                container_port=5000
            )
        ],
        health_check_grace_period_seconds=60,
        tags={
            "Name": f"flask-service-{environment_suffix}",
            "EnvironmentSuffix": environment_suffix
        },
        opts=pulumi.ResourceOptions(depends_on=depends_on_resources)
    )

    return {
        "service": service,
        "task_definition": task_definition,
        "task_execution_role": task_execution_role,
        "task_role": task_role,
        "log_group": log_group
    }

File: archive/pulumi-ts/Pr7161/lib/components/ecs.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface EcsComponentArgs {
  environment: string;
  clusterArn: pulumi.Output<string>;
  vpcId: pulumi.Output<string>;
  privateSubnetIds: pulumi.Output<string>[];
  targetGroupArn: pulumi.Output<string>;
  ecrRepositoryUrl: pulumi.Output<string>;
  dbSecretArn: pulumi.Output<string>;
  scalingCpuThreshold: number;
  albSecurityGroupId: pulumi.Output<string>;
  tags: { [key: string]: string };
}

export class EcsComponent extends pulumi.ComponentResource {
  public readonly taskDefinition: aws.ecs.TaskDefinition;
  public readonly service: aws.ecs.Service;
  public readonly taskRole: aws.iam.Role;
  public readonly executionRole: aws.iam.Role;
  public readonly taskExecutionRole: aws.iam.Role; // Alias for executionRole for test compatibility
  public readonly logGroup: aws.cloudwatch.LogGroup;
  public readonly scalingTarget: aws.appautoscaling.Target;
  public readonly scalingPolicy: aws.appautoscaling.Policy;
  private readonly securityGroup: aws.ec2.SecurityGroup;

  constructor(
    name: string,
    args: EcsComponentArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('custom:payment:EcsComponent', name, {}, opts);

    const resourceOpts = { parent: this };

    // Create security group for ECS tasks
    this.securityGroup = new aws.ec2.SecurityGroup(
      `${args.environment}-payment-ecs-sg`,
      {
        name: `${args.environment}-payment-ecs-sg`,
        description: 'Security group for ECS tasks',
        vpcId: args.vpcId,
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 8080,
            toPort: 8080,
            securityGroups: [args.albSecurityGroupId],
            description: 'Allow traffic from ALB',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          ...args.tags,
          Name: `${args.environment}-payment-ecs-sg`,
        },
      },
      resourceOpts
    );

    // Create ECS task execution role
    this.executionRole = new aws.iam.Role(
      `${args.environment}-payment-ecs-execution-role`,
      {
        name: `${args.environment}-payment-ecs-execution-role`,
        assumeRolePolicy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Action: 'sts:AssumeRole',
              Effect: 'Allow',
              Principal: {
                Service: 'ecs-tasks.amazonaws.com',
              },
            },
          ],
        }),
        tags: args.tags,
      },
      resourceOpts
    );

    // Attach execution role policies
    new aws.iam.RolePolicyAttachment(
      `${args.environment}-payment-ecs-execution-policy`,
      {
        role: this.executionRole.name,
        policyArn:
          'arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy',
      },
      resourceOpts
    );

    // Additional policy for Secrets Manager access
    const secretsPolicy = new aws.iam.Policy(
      `${args.environment}-payment-secrets-policy`,
      {
        name: `${args.environment}-payment-secrets-policy`,
        policy: args.dbSecretArn.apply(arn =>
          JSON.stringify({
            Version: '2012-10-17',
            Statement: [
              {
                Effect: 'Allow',
                Action: [
                  'secretsmanager:GetSecretValue',
                  'secretsmanager:DescribeSecret',
                ],
                Resource: arn,
              },
            ],
          })
        ),
        tags: args.tags,
      },
      resourceOpts
    );

    new aws.iam.RolePolicyAttachment(
      `${args.environment}-payment-secrets-attachment`,
      {
        role: this.executionRole.name,
        policyArn: secretsPolicy.arn,
      },
      resourceOpts
    );

    // Create ECS task role
    this.taskRole = new aws.iam.Role(
      `${args.environment}-payment-ecs-task-role`,
      {
        name: `${args.environment}-payment-ecs-task-role`,
        assumeRolePolicy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Action: 'sts:AssumeRole',
              Effect: 'Allow',
              Principal: {
                Service: 'ecs-tasks.amazonaws.com',
              },
            },
          ],
        }),
        tags: args.tags,
      },
      resourceOpts
    );

    // Create CloudWatch log group (name must follow AWS naming conventions)
    const logGroupName =
      `/ecs/${args.environment}-payment-processor`.toLowerCase();
    this.logGroup = new aws.cloudwatch.LogGroup(
      `${args.environment}-payment-logs`,
      {
        name: logGroupName,
        retentionInDays: 7,
        tags: args.tags,
      },
      resourceOpts
    );

    // Create ECS task definition (family name must be lowercase)
    const taskFamily = `${args.environment}-payment-processor`.toLowerCase();
    this.taskDefinition = new aws.ecs.TaskDefinition(
      `${args.environment}-payment-task`,
      {
        family: taskFamily,
        networkMode: 'awsvpc',
        requiresCompatibilities: ['FARGATE'],
        cpu: '256',
        memory: '512',
        executionRoleArn: this.executionRole.arn,
        taskRoleArn: this.taskRole.arn,
        containerDefinitions: pulumi
          .all([args.ecrRepositoryUrl, args.dbSecretArn])
          .apply(([repoUrl, secretArn]) =>
            JSON.stringify([
              {
                name: 'payment-processor',
                image: `${repoUrl}:latest`,
                essential: true,
                portMappings: [
                  {
                    containerPort: 8080,
                    protocol: 'tcp',
                  },
                ],
                environment: [
                  { name: 'ENVIRONMENT', value: args.environment },
                  { name: 'PORT', value: '8080' },
                ],
                secrets: [
                  {
                    name: 'DATABASE_CREDENTIALS',
                    valueFrom: secretArn,
                  },
                ],
                logConfiguration: {
                  logDriver: 'awslogs',
                  options: {
                    'awslogs-group': logGroupName,
                    'awslogs-region': aws.getRegionOutput().name,
                    'awslogs-stream-prefix': 'ecs',
                  },
                },
                healthCheck: {
                  command: [
                    'CMD-SHELL',
                    'curl -f http://localhost:8080/health || exit 1',
                  ],
                  interval: 30,
                  timeout: 5,
                  retries: 3,
                  startPeriod: 60,
                },
              },
            ])
          ),
        tags: args.tags,
      },
      resourceOpts
    );

    // Create ECS service
    this.service = new aws.ecs.Service(
      `${args.environment}-payment-service`,
      {
        name: `${args.environment}-payment-service`,
        cluster: args.clusterArn,
        taskDefinition: this.taskDefinition.arn,
        desiredCount: 2,
        launchType: 'FARGATE',
        platformVersion: 'LATEST',
        networkConfiguration: {
          assignPublicIp: false,
          subnets: args.privateSubnetIds,
          securityGroups: [this.securityGroup.id],
        },
        loadBalancers: [
          {
            targetGroupArn: args.targetGroupArn,
            containerName: 'payment-processor',
            containerPort: 8080,
          },
        ],
        healthCheckGracePeriodSeconds: 60,
        enableExecuteCommand: true,
        tags: args.tags,
      },
      resourceOpts
    );

    // Create auto-scaling target
    this.scalingTarget = new aws.appautoscaling.Target(
      `${args.environment}-payment-scaling-target`,
      {
        maxCapacity: 10,
        minCapacity: 2,
        resourceId: pulumi.interpolate`service/${args.clusterArn.apply(arn => arn.split('/')[1])}/${this.service.name}`,
        scalableDimension: 'ecs:service:DesiredCount',
        serviceNamespace: 'ecs',
      },
      resourceOpts
    );

    // Create auto-scaling policy
    this.scalingPolicy = new aws.appautoscaling.Policy(
      `${args.environment}-payment-scaling-policy`,
      {
        name: `${args.environment}-payment-cpu-scaling`,
        policyType: 'TargetTrackingScaling',
        resourceId: this.scalingTarget.resourceId,
        scalableDimension: this.scalingTarget.scalableDimension,
        serviceNamespace: this.scalingTarget.serviceNamespace,
        targetTrackingScalingPolicyConfiguration: {
          targetValue: args.scalingCpuThreshold,
          predefinedMetricSpecification: {
            predefinedMetricType: 'ECSServiceAverageCPUUtilization',
          },
          scaleInCooldown: 300,
          scaleOutCooldown: 60,
        },
      },
      resourceOpts
    );

    // Alias for test compatibility
    this.taskExecutionRole = this.executionRole;

    this.registerOutputs({
      taskDefinitionArn: this.taskDefinition.arn,
      serviceArn: this.service.id,
    });
  }

  // Helper method to get security group ID
  public getSecurityGroupId(): pulumi.Output<string> {
    return this.securityGroup.id;
  }
}
